syntax = "proto3";

option java_multiple_files = true;
option java_package = "org.example.ticket";
option java_outer_classname = "EventTicket";

package ticket;

import "google/api/annotations.proto";

service TicketService {
    rpc BookTicket (EventRequest) returns (Ticket) {
        option (google.api.http) = {
            post: "/ticket/create"
            body: "*"
          };
    }
    rpc CancelTicket (TicketRequest) returns (TicketResponse) {
        option (google.api.http) = {
            post: "/ticket/cancel"
            body: "*"
          };
    }
    rpc GetTicket (TicketRequest) returns (Ticket) {
        option (google.api.http) = {
            post: "/ticket/get"
            body: "*"
          };
    }
    rpc GetTickets (EventRequest) returns (stream Ticket) {}
}

message EventRequest {
    string event_id = 1;
    string user_id = 2;
}

message TicketRequest {
    string ticket_id = 1;
    string user_id = 2;
}

message TicketResponse {
    string message = 1;
}

message Ticket {
    string ticket_id = 1;
    string user_id = 2;
    string event_id = 3;
    int64 date_purchased = 4;
    enum StatusCode {
        PENDING = 0;
        BOOKED = 1;
        CANCELLED = 2;
        INVALID = 3;
    }
    StatusCode status = 5;
}